modules/service/variables.tf

```diff
@@ -0,0 +1,638 @@
+}
+
+################################################################################
+# Service - IAM Role
+################################################################################
+
+variable "create_iam_role" {
+  description = "Determines whether the ECS service IAM role should be created"
+  type        = bool
+  default     = true
+}
+
+variable "iam_role_arn" {
+  description = "Existing IAM role ARN"
+  type        = string
+  default     = null
+}
+
+variable "iam_role_name" {
+  description = "Name to use on IAM role created"
+  type        = string
+  default     = null
+}
+
+variable "iam_role_use_name_prefix" {
+  description = "Determines whether the IAM role name (`iam_role_name`) is used as a prefix"
+  type        = bool
+  default     = true
+}
+
+variable "iam_role_path" {
+  description = "IAM role path"
+  type        = string
+  default     = null
+}
+
+variable "iam_role_description" {
+  description = "Description of the role"
+  type        = string
+  default     = null
+}
+
+variable "iam_role_permissions_boundary" {
+  description = "ARN of the policy that is used to set the permissions boundary for the IAM role"
+  type        = string
+  default     = null
+}
+
+variable "iam_role_tags" {
+  description = "A map of additional tags to add to the IAM role created"
+  type        = map(string)
+  default     = {}
+}
+
+variable "iam_role_statements" {
+  description = "A map of IAM policy [statements](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document#statement) for custom permission usage"
+  type        = any
+  default     = {}
+}
+
+################################################################################
+# Task Definition
+################################################################################
+
+variable "create_task_definition" {
+  description = "Determines whether to create a task definition or use existing/provided"
+  type        = bool
+  default     = true
+}
+
+variable "task_definition_arn" {
+  description = "Existing task definition ARN. Required when `create_task_definition` is `false`"
+  type        = string
+  default     = null
+}
+
+variable "container_definitions" {
+  description = "A map of valid [container definitions](http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html). Please note that you should only provide values that are part of the container definition document"
+  type        = any
+  default     = {}
+}
+
+variable "container_definition_defaults" {
+  description = "A map of default values for [container definitions](http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html) created by `container_definitions`"
+  type        = any
+  default     = {}
+}
+
+variable "cpu" {
+  description = "Number of cpu units used by the task. If the `requires_compatibilities` is `FARGATE` this field is required"
+  type        = number
+  default     = 1024
```

Comment:
See latter part of the variable description - since the default is `FARGATE` (most common), then this value is required so we provide a default value to satisfy

Reply:
its the pattern you typically see with containerized based deployments. the concurrency model shifts from a larger resource with multi-threading and/or parallelism within one pod/container to instead use parallelism across multiple pods/containers. this allows you to scope your resource utilization for "one unit of work" to a pod and then the container orchestrator knows how to scale that based on the workload demand. this allows you to optimize resource utilization efficiently - imagine having pods that were using 4096 for vCPU (4 virtual cores) you have to scale in much larger steps which means you will have periods where you are not fully utilizing the resources provisioned. you also have to factor this into your autoscaling policy - its no longer a 1:1 but a 1:4, your code is more complex if you opt into multi-threading and/or parallelism in order to fully utilize the 4 virtual cores, etc.